File: packages/js/core/src/team.js
import { DEFAULT_PROTOCOL_MODE, agentName, toJSONString } from "./utils.js";
import { HandoffState } from "./contracts.js";
import { AgentRunResult } from "./agent.js";

export class HandoffProtocol {
  constructor({ mode = DEFAULT_PROTOCOL_MODE } = {}) {
    this.mode = mode;
  }

  transfer({ fromAgent, toAgent, task, summary, decisions = [], importantFiles = [], errors = [], nextSteps = [], contextRefs = [], metadata = {} }) {
    return new HandoffState({
      task,
      fromAgent: agentName(fromAgent),
      toAgent: agentName(toAgent),
      summary,
      decisions,
      importantFiles,
      errors,
      nextSteps,
      contextRefs,
      metadata: { protocolMode: this.mode, ...metadata },
    }).validate();
  }
}

export class Team {
  constructor({ agents = [], protocol = new HandoffProtocol(), metadata = {} } = {}) {
    const list = Array.isArray(agents) ? agents : [];
    if (list.length === 0) {
      throw new TypeError("Team requires at least one agent.");
    }
    this.agents = [...list];
    this.protocol = protocol;
    this.metadata = metadata ? { ...metadata } : {};
  }

  run(task) {
    const stepResults = [];
    const handoffs = [];
    let currentTask = task;
    let previousAgent = null;
    let previousResult = null;

    for (const agent of this.agents) {
      const result = agent.run(currentTask, {
        context: previousResult?.finalOutput ?? null,
      });
      stepResults.push(result);

      if (previousAgent) {
        handoffs.push(
          this.protocol.transfer({
            fromAgent: previousAgent,
            toAgent: agent,
            task: currentTask,
            summary: previousResult.finalOutput,
            decisions: [`${previousAgent.name} completed its step.`],
            nextSteps: [`${agent.name} should continue from structured state.`],
          }),
        );
      }

      previousAgent = agent;
      previousResult = result;
      currentTask = result.finalOutput;
    }

    return new TeamRunResult({
      success: stepResults.every((result) => result.success),
      task,
      finalOutput: stepResults.at(-1)?.finalOutput ?? "",
      stepResults,
      handoffs,
      metadata: this.metadata,
    });
  }

  async arun(task) {
    const stepResults = [];
    const handoffs = [];
    let currentTask = task;
    let previousAgent = null;
    let previousResult = null;

    for (const agent of this.agents) {
      const result = await agent.arun(currentTask, {
        context: previousResult?.finalOutput ?? null,
      });
      stepResults.push(result);

      if (previousAgent) {
        handoffs.push(
          this.protocol.transfer({
            fromAgent: previousAgent,
            toAgent: agent,
            task: currentTask,
            summary: previousResult.finalOutput,
            decisions: [`${previousAgent.name} completed its step.`],
            nextSteps: [`${agent.name} should continue from structured state.`],
          }),
        );
      }

      previousAgent = agent;
      previousResult = result;
      currentTask = result.finalOutput;
    }

    return new TeamRunResult({
      success: stepResults.every((result) => result.success),
      task,
      finalOutput: stepResults.at(-1)?.finalOutput ?? "",
      stepResults,
      handoffs,
      metadata: this.metadata,
    });
  }
}

export class TeamRunResult {
  constructor({ success, task, finalOutput, stepResults = [], handoffs = [], metadata = {} }) {
    this.success = Boolean(success);
    this.task = task;
    this.finalOutput = finalOutput;
    this.stepResults = Array.isArray(stepResults) ? [...stepResults] : [];
    this.handoffs = Array.isArray(handoffs) ? [...handoffs] : [];
    this.metadata = metadata ? { ...metadata } : {};
  }

  toJSON() {
    return {
      success: this.success,
      task: this.task,
      final_output: this.finalOutput,
      step_results: this.stepResults.map((result) => result.toJSON()),
      handoffs: this.handoffs.map((handoff) => handoff.toJSON()),
      metadata: { ...this.metadata },
    };
  }

  toJSONString(space = 2) {
    return JSON.stringify(this.toJSON(), null, space);
  }

  static fromJSON(value) {
    const data = typeof value === "string" ? JSON.parse(value) : value;
    return new TeamRunResult({
      success: data.success,
      task: data.task,
      finalOutput: data.finalOutput ?? data.final_output,
      stepResults: (Array.isArray(data.stepResults ?? data.step_results) ? data.stepResults ?? data.step_results : [])
        .map((result) => result instanceof AgentRunResult ? result : AgentRunResult.fromJSON(result)),
      handoffs: (Array.isArray(data.handoffs) ? data.handoffs : [])
        .map((handoff) => handoff instanceof HandoffState ? handoff : HandoffState.fromJSON(handoff)),
      metadata: data.metadata,
    });
  }
}

File: cpp/packages/handoffkit/src/runtime/team.cpp
#include <handoffkit/runtime/team.hpp>

namespace handoffkit {

Result<TeamRunResult> Team::run(std::string_view task) {
    if (agents_.empty()) {
        return Error::invalid_argument("Team requires at least one agent.", "agents");
    }
    if (task.empty()) {
        return Error::invalid_argument("Task cannot be empty.", "task");
    }

    TeamRunResult result;
    result.task = std::string(task);

    Agent& first = agents_.front();
    auto first_out = first.run(task);
    if (!first_out) {
        return first_out.error();
    }
    std::string current_output = std::move(first_out.value());
    result.agent_outputs.push_back(AgentOutput{first.name(), current_output});

    for (std::size_t i = 1; i < agents_.size(); ++i) {
        Agent& previous = agents_[i - 1];
        Agent& current = agents_[i];

        TransferOptions transfer;
        transfer.task = result.task;
        transfer.from_agent = previous.name();
        transfer.to_agent = current.name();
        transfer.summary = current_output;

        auto handoff = protocol_.transfer(transfer);
        if (!handoff) {
            return handoff.error();
        }
        result.handoffs.push_back(handoff.value());

        RunOptions opts;
        opts.handoff = &result.handoffs.back();
        auto step = current.run(task, opts);
        if (!step) {
            return step.error();
        }
        current_output = std::move(step.value());
        result.agent_outputs.push_back(AgentOutput{current.name(), current_output});
    }

    result.final_output = std::move(current_output);
    result.success = true;
    return Result<TeamRunResult>::success(std::move(result));
}

}  // namespace handoffkit
